source/common/stats/symbol_table_impl.cc

```diff
@@ -51,45 +67,59 @@ void SymbolTableImpl::Encoding::addSymbol(Symbol symbol) {
   do {
-    if (symbol < (1 << 7)) {
-      vec_.push_back(symbol); // symbols <= 127 get encoded in one byte.
+    if (number < (1 << 7)) {
+      *bytes++ = number; // number <= 127 get encoded in one byte.
     } else {
-      vec_.push_back((symbol & Low7Bits) | SpilloverMask); // symbols >= 128 need spillover bytes.
+      *bytes++ = ((number & Low7Bits) | SpilloverMask); // number >= 128 need spillover bytes.
     }
-    symbol >>= 7;
-  } while (symbol != 0);
+    number >>= 7;
+  } while (number != 0);
+  return bytes;
+}
+
+void SymbolTableImpl::Encoding::addSymbols(const std::vector<Symbol>& symbols) {
+  ASSERT(data_bytes_required_ == 0);
+  for (Symbol symbol : symbols) {
+    data_bytes_required_ += encodingSizeBytes(symbol);
+  }
+  storage_ = std::make_unique<Storage>(data_bytes_required_);
+  uint8_t* bytes = storage_.get();
+  for (Symbol symbol : symbols) {
+    bytes = writeEncodingReturningNext(symbol, bytes);
+  }
+  ASSERT(static_cast<uint64_t>(bytes - storage_.get()) == data_bytes_required_);
+}
+
+std::pair<uint64_t, uint64_t> SymbolTableImpl::Encoding::decodeNumber(const uint8_t* encoding) {
+  uint64_t number = 0;
+  uint64_t uc = SpilloverMask;
+  const uint8_t* start = encoding;
+  for (uint32_t shift = 0; (uc & SpilloverMask) != 0; ++encoding, shift += 7) {
+    uc = static_cast<uint32_t>(*encoding);
+    number |= (uc & Low7Bits) << shift;
+  }
+  return std::make_pair(number, encoding - start);
 }
 
 SymbolVec SymbolTableImpl::Encoding::decodeSymbols(const SymbolTable::Storage array,
                                                    uint64_t size) {
   SymbolVec symbol_vec;
-  Symbol symbol = 0;
-  for (uint32_t shift = 0; size > 0; --size, ++array) {
-    uint32_t uc = static_cast<uint32_t>(*array);
-
-    // Inverse addSymbol encoding, walking down the bytes, shifting them into
-    // symbol, until a byte with a zero high order bit indicates this symbol is
-    // complete and we can move to the next one.
-    symbol |= (uc & Low7Bits) << shift;
-    if ((uc & SpilloverMask) == 0) {
-      symbol_vec.push_back(symbol);
-      shift = 0;
-      symbol = 0;
-    } else {
-      shift += 7;
-    }
+  symbol_vec.reserve(size);
+  while (size > 0) {
+    std::pair<uint64_t, uint64_t> symbol_consumed = decodeNumber(array);
+    symbol_vec.push_back(symbol_consumed.first);
+    size -= symbol_consumed.second;
+    array += symbol_consumed.second;
   }
   return symbol_vec;
 }
 
-uint64_t SymbolTableImpl::Encoding::moveToStorage(SymbolTable::Storage symbol_array) {
-  const uint64_t sz = dataBytesRequired();
-  symbol_array = writeLengthReturningNext(sz, symbol_array);
-  if (sz != 0) {
-    memcpy(symbol_array, vec_.data(), sz * sizeof(uint8_t));
+void SymbolTableImpl::Encoding::moveToStorage(SymbolTable::Storage symbol_array) {
+  symbol_array = writeEncodingReturningNext(data_bytes_required_, symbol_array);
+  if (data_bytes_required_ != 0) {
+    memcpy(symbol_array, storage_.get(), data_bytes_required_);
```

Comment:
Is it possible to eliminate `memcpy` in this file? I would prefer we did if there is a slower but less dangerous way to do this. I.e. can we use `std::vector` or some other STL structure?

Reply:
Speed is not the issue; it's overhead on a structure that gets used many times.

I'm definitely all ears if there's ideas to structure this code in a way that makes it more locally obvious we are not writing out-of-bounds here.

Note there are two uses of moveToStorage: a simple one in `SymbolTableImpl::encode` where we allocate storage for only one encoding, and then move into that storage. If I inlined this into that call-site it would be reasonably obvious it was correct.

But a significant memory win was in the usage of this at another call-site, in `SymbolTableImpl::populateList` where we do a single allocation covering the original stat-name, the tag extracted name, and all tags and values.

I'll also note that this PR does not add any memcpy; it's only that a nearby refactor touched the line.

One of the benefits of his PR -- actually the main motivation -- was to help the fuzzer hit more cases.

Reply:
@jmarantz were you planning on implementing the approach you discussed in IM? In general, this PR looks OK, but it's all really complicated and I'd like a second set of eyes on whatever we end up with here. Basically, this is low level memory management code, and I think you have a clear justification to be doing memcpy etc., but we should have higher scrutiny of code like this than other PRs.

Reply:
@htuch @asraa yes, this is working now but needs cleanup, more unit tests, and actually we might be able to fuzz test the new source/common/common/mem_block.h directly.

The memcpy() in there I think is locally easy it inspect as writing in-bounds, based on the asserts. It's a little tougher to guarantee the provenance of the bytes it's reading *from*. Interested in comments you may have (beyond adding tests, comments, & fuzz).

Reply:
Resolving this comment now as the raw calls to memcpy() have been eliminated in favor of MemBlockBuilder.